allennlp/nn/util.py

```diff
@@ -344,23 +342,24 @@ def sequence_cross_entropy_with_logits(logits: torch.FloatTensor,
     if batch_average:
         num_non_empty_sequences = ((weights.sum(1) > 0).float().sum() + 1e-13)
         return per_batch_loss.sum() / num_non_empty_sequences
-    return per_batch_loss.squeeze(1)
+    return per_batch_loss
 
 
 def replace_masked_values(tensor: Variable, mask: Variable, replace_with: float) -> Variable:
     """
     Replaces all masked values in ``tensor`` with ``replace_with``.  ``mask`` must be broadcastable
     to the same shape as ``tensor``.
-
-    Until pytorch 0.2 is released, we require that ``tensor.dim() == mask.dim()``.  Otherwise we
-    won't know which dimensions of the mask to unsqueeze.
     """
-    # We'll build a tensor of the same shape as `tensor`, subtract away masked values, then add
-    # back in the `replace_with` value.
-    # TODO(mattg): use broadcasting here when it's available in pytorch 0.2.
-    if tensor.dim() != mask.dim():
-        raise ConfigurationError("tensor.dim() (%d) != mask.dim() (%d)" % (tensor.dim(), mask.dim()))
-    one_minus_mask = 1.0 - mask.expand_as(tensor)
+    # We'll build a tensor of the same shape as `tensor`, subtract away
+    # masked values, then add back in the `replace_with` value.
+    if tensor.dim() < mask.dim():
+        raise ConfigurationError("Mask has dimension {} but tensor has dimension {}. Mask dimension "
+                                 "should be <= tensor dimension.".format(mask.dim(), tensor.dim()))
+    # Expand mask dimensions to match tensor if it's smaller.
+    while mask.dim() < tensor.dim():
+        mask = mask.unsqueeze(-1)
```

Comment:
I'm actually not sure we can do this safely.  I think my original comment about requiring `tensor.dim() == mask.dim()` only until pytorch 0.2 was wrong, and we probably should just leave this as it was (except we can probably remove the `expand_as`).

Consider:

If this is for an embedding, we have tensor with shape `(batch_size, num_tokens, embedding_dim)`, with mask `(batch_size, num_tokens)`.  We want to unsqueeze the last dimension, and that's fine.  But in BiDAF, I use this with a tensor of shape `(batch_size, passage_length, question_length)`, with a mask of shape `(batch_size, question_length)`.  Unsqueezing the last dimension here is wrong, and could lead to odd and unpredictable behavior of `passage_length` happens to equal `question_length` - the mask will be applied on the wrong dimension.  I think we just need to require the caller to have put the mask into the same dimension as the tensor, so we don't have to guess in this function.

I'm a little concerned that this passes the BiDAF test, actually - I'm not sure how this passes.